extensions/2.0/Khronos/KHR_materials_volume/README.md

```diff
@@ -0,0 +1,276 @@
+</figure>
+
+The volume extension has to be combined with `KHR_materials_transmission` or `KHR_materials_translucency`. Light that falls onto the volume boundary may enter the volume, depending on the transmission or translucency of the surface's BSDF. Inside the volume, light might be absorbed or scattered by particles in the medium, and eventually hit some surface from inside the volume. It can be a different surface and will most probably be a different point in space compared to the entrance point. At this exit point, if the material properties allow, light may leave the volume.
+
+<figure style="text-align:center">
+<img src="./figures/volume.svg"/>
+<figcaption><em>Interaction of light rays inside the volume and at the boundaries. The volume is homogeneous and has an index of refraction of 1.5.</figcaption>
+</em></figure>
+
+When light interacts with the surface, it is reflected or refracted at microfacets, taking the roughness of the material into account. The index of refraction is taken from `KHR_materials_ior`.
+
+## Extending Materials
+
+The volumetric properties are defined by adding the `KHR_materials_volume` extension to any glTF material. A non-zero thickness switches from thin-walled to volumetric behavior. This requires a manifold/closed mesh. The properties of the medium are not texturable, it is assumed to be homogeneous.
+
+```json
+materials: [
+    {
+        "extensions": {
+            "KHR_materials_volume": {
+                "thicknessFactor": 1.0,
+                "attenuationDistance":  0.006,
+                "attenuationColor": [ 0.5, 0.5, 0.5 ],
+                "subsurfaceColor": [ 0.572, 0.227, 0.075 ]
+            }
+        }
+    }
+]
+```
+
+## Properties
+
+The extension defines the following parameters to describe the volume.
+
+| | Type | Description | Required |
+|-|------|-------------|----------|
+| **thicknessFactor** | `number` | The thickness of the volume beneath the surface in meters in the local coordinate system of the node. If the value is 0 the material is thin-walled. Otherwise the material is a volume boundary. The `doubleSided` property has no effect on volume boundaries. | No, default: `0`. |
+| **thicknessTexture** | `textureInfo` | A texture that defines the thickness, stored in the G channel. This will be multiplied by `thicknessFactor`. | No |
+| **attenuationDistance** | `number` | Density of the medium given as the average distance in meters that light travels in the medium before interacting with a particle. | No, default: +Infinity |
+| **attenuationColor** | `number[3]` | The color that white light turns into due to absorption when reaching the attenuation distance. | No, default: `[1, 1, 1]` |
+| **subsurfaceColor** | `number[3]` | The overall color perceived at the surface due to subsurface scattering. | No, default: `[0, 0, 0]` |
```

Comment:
I would propose to remove _subsurfaceColor_ and rely on _baseColor_ - mainly for two reasons:
- I don't think there is a realistic usecase where we need both _baseColor_ AND _subsurfaceColor_
- It could be argued that we are already (sort of) modelling this effect with the diffuse contribution.

I think it would simplify the solution to only have _attenuationDistance_ and _attenuationColor_ without sacrificing material representation.

Reply:
To follow up on yesterdays discussion on the pbr tsg call:
I am happy with the naming as it is since it is a combination of several physical properties that we do not need to model separately (scattering + absorption)

Reply:
I would also like to add a non linear factor for the absorption rate (attenuation coefficient) - a lot of our translucent materials use this.
What I mean by this is an exponential function for the attenuation as the light interacts with particles within the volume - a value of 1.0 means linear attenuation, larger values will make the light absorb faster and vice versa.

Reply:
> I would propose to remove subsurfaceColor and rely on baseColor [...]

There are a lot of similarities between *subsurfaceColor* and *baseColor*, but there is one important difference: *baseColor* can be textured (2D), but *subsurfaceColor* must be constant (or a 3D texture). Thus we need to split it up.

> I would also like to add a non linear factor for the absorption rate (attenuation coefficient)

Do you mean providing an *attenuationCoefficient* parameter instead of *attenuationColor* and *attenuationDistance*? Note that *attenuationCoefficient* = -log(*attenuationColor*)/*attenuationDistance*.

Reply:
> There are a lot of similarities between _subsurfaceColor_ and _baseColor_, but there is one important difference: _baseColor_ can be textured (2D), but _subsurfaceColor_ must be constant (or a 3D texture). Thus we need to split it up.

Hmm, I don't think I understand the reasoning for this.
As defined in the transmission extension the baseColor is used to tint the light that enters through the medium.
I think the same should be the case for this extension - or am I missing something?

Conceptually, I see it something like this:
Light enters through the medium and is colored by the baseColor, it travels _attenuationDistance_ before interacting with a particle (inside the volume).
If light escapes the volume after interacting with a particle the entered light will be colored by the attenuationColor and affected by the attenuationDistance.
If light escapes the volume without interacting with a particle it will be the same as the entered light.

Is this in line with what you have in mind?

Reply:
> Do you mean providing an _attenuationCoefficient_ parameter instead of _attenuationColor_ and _attenuationDistance_? Note that _attenuationCoefficient_ = -log(_attenuationColor_)/_attenuationDistance_.

I think so - sorry for not grasping all of your calculations correctly :-)

The effect I am after is a non-linear falloff of escaping light based on the object thickness - as the above calculations does not take thickness into account I cannot see how that would be done.....

Reply:
Hi @rsahlin. This extension is all about thickness. The parameters combine to do exactly what you're describing. The light is attenuated based on thickness. The attenuation coefficient is calculated as @proog128 said but it is applied using the thickness and Beer's Law.
`baseColor` is a constant absorption only at the surface.

Reply:
Again, sorry for not being able to fully follow your calculations.

I think that I need to spend some time understanding this and how it relates to the properties we want our materials to exhibit....